r/src/array.cpp

```diff
@@ -828,36 +849,21 @@ SEXP TimeArray_to_Vector(const std::shared_ptr<Array>& array, int32_t multiplier
 
 SEXP DecimalArray(const std::shared_ptr<Array>& array) {
   auto n = array->length();
-  NumericVector vec(no_init(n));
-
-  if (n == 0) return vec;
-
-  auto null_count = array->null_count();
-  if (null_count == n) {
-    std::fill(vec.begin(), vec.end(), NA_REAL);
-    return vec;
-  }
+  ComplexVector vec(no_init(n));
 
-  auto p_vec = reinterpret_cast<double*>(vec.begin());
-  const auto& decimals_arr =
-      internal::checked_cast<const arrow::Decimal128Array&>(*array);
-
-  if (array->null_count()) {
-    internal::BitmapReader bitmap_reader(array->null_bitmap()->data(), array->offset(),
-                                         n);
+  auto p_vec = reinterpret_cast<arrow::Decimal128*>(vec.begin());
+  auto p_values = GetValuesSafely<arrow::Decimal128>(array->data(), 1, array->offset());
 
-    for (size_t i = 0; i < n; i++, bitmap_reader.Next()) {
-      p_vec[i] = bitmap_reader.IsNotSet()
-                     ? NA_REAL
-                     : std::stod(decimals_arr.FormatValue(i).c_str());
-    }
-  } else {
-    for (size_t i = 0; i < n; i++) {
-      p_vec[i] = std::stod(decimals_arr.FormatValue(i).c_str());
-    }
-  }
+  // TODO: deal with NA
+  std::copy_n(p_values, n, p_vec);
```

Comment:
This looks fishy. See

```
void Decimal128::ToBytes(uint8_t* out) const {
  DCHECK_NE(out, nullptr);
  reinterpret_cast<uint64_t*>(out)[0] = BitUtil::ToLittleEndian(low_bits_);
  reinterpret_cast<int64_t*>(out)[1] = BitUtil::ToLittleEndian(high_bits_);
}
```

but in `Decimal128` we have

```
 private:
  int64_t high_bits_;
  uint64_t low_bits_;
```

In principle I don't see why these could not be flipped. @cpcloud?

Reply:
There are number of things that I don't think will work here.

1. `ComplexVector` I'm assuming is a sequence of pairs of `double`s which while in principle you could use as a surrogate for any old 128 bit thing, is very confusing to read IMO and loses precision as soon as you start assuming the `int64_t`/`uint64_t` pair fit inside of a `double`.
1. I'm not exactly sure how C++ classes are laid out in memory, but I don't think you can guarantee where the compiler will place data members. In this case, you have the reverse problem where you are trying to view a pair of doubles as a class. I don't know what to expect here, but I'd be pretty surprised if there were any guarantees about what you'd be able to with the `Decimal128`s you're `reinterpret_cast`ing to. For example, can you call a method on these objects?
1. Finally, If you're just using the `ComplexVector` for storage then I'm wondering if there was a better way. Maybe two int64 arrays? I'm not familiar with how such a thing would work in R or if that's somehow more difficult than the current approach.

I think it will be easiest to use a simple loop to assign the `Decimal128Array` values to whatever the output array ends up being, and access the high and low bits of each `Decimal128` value through the provided APIs. This would, for example, insulate your code against any private details of the `Decimal128` implementation that may change in the future. Not sure what those changes would be necessarily, but you don't want to be exposed to them if you don't need to be.

Reply:
On 

1. AFAICT ComplexVector is being used here as a convenience since it's 16 bytes wide and already available in Rcpp

2. I think you can assume the same memory order as a C struct. 

3. see point 1 :)

Reply:
> I think you can assume the same memory order as a C struct.

What about for classes with virtual methods?

Reply:
Ah, this only applies to POD-like classes. If you have a base class with data members, then subclasses have the base members first in their memory layout

so

```
struct Foo {
  int32_t a;
  int32_t b;
};

struct Bar : public Foo {
  int32_t c;
};
```

Then the first 8 bytes of a Bar is the Foo members

Reply:
The vtable isn't part of the object's memory layout; calls into the vtable are dealt with during compilation, and the vtable is stored elsewhere

Reply:
In this particular case, there aren't any so it should be okay.

Reply:
yeah complex vectors are just used here because they are 16 bytes. The R code around it using the `vctrs` package makes sure we never care about each `double`. As far as this is concerned it is just a contiguous sequence of 16 bytes.